From 7358eafaedce1a9d8ced794a4c4b272f2a516ce0 Mon Sep 17 00:00:00 2001
From: Michel JAOUEN <michel.jaouen@st.com>
Date: Mon, 20 Nov 2023 11:30:11 +0100
Subject: [PATCH 0852/1695] drivers: stm32_ipcc: add mailbox support

Notification between M33 and optee is required for the support a
new scmi client in m33. ipcc drivers ipcc is updated for this
support.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git bba315eb43821dedded26f774c2d9e9f3a85797c drivers: stm32_ipcc: add mailbox support]
Change-Id: I855528e399d3eae6be5d6dac6db05a3df62f6f1b
Signed-off-by: Michel JAOUEN <michel.jaouen@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/378806
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
---
 core/drivers/stm32_ipcc.c | 388 +++++++++++++++++++++++++++++++++++++-
 1 file changed, 386 insertions(+), 2 deletions(-)

diff --git a/core/drivers/stm32_ipcc.c b/core/drivers/stm32_ipcc.c
index 7921540fd..a4023af1d 100644
--- a/core/drivers/stm32_ipcc.c
+++ b/core/drivers/stm32_ipcc.c
@@ -1,6 +1,6 @@
 // SPDX-License-Identifier: BSD-3-Clause
 /*
- * Copyright (c) 2022, STMicroelectronics
+ * Copyright (C) 2022-2024, STMicroelectronics
  */
 
 #include <arm.h>
@@ -8,12 +8,15 @@
 #include <drivers/clk.h>
 #include <drivers/clk_dt.h>
 #include <drivers/stm32_rif.h>
+#include <drivers/mailbox.h>
 #include <kernel/boot.h>
 #include <kernel/delay.h>
 #include <kernel/dt.h>
 #include <kernel/dt_driver.h>
+#include <kernel/interrupt.h>
 #include <kernel/panic.h>
 #include <kernel/pm.h>
+#include <kernel/spinlock.h>
 #include <libfdt.h>
 #include <mm/core_memprot.h>
 #include <stdbool.h>
@@ -29,6 +32,43 @@
 #define IPCC_C2CIDCFGR			U(0x98)
 #define IPCC_HWCFGR			U(0x3F0)
 
+#define IPCC_C1CR			U(0x00)
+#define IPCC_C1MR			U(0x04)
+#define IPCC_C1SCR			U(0x08)
+#define IPPC_C1TOC2SR			U(0x0C)
+
+#define IPCC_C2CR			U(0x10)
+#define IPCC_C2MR			U(0x14)
+#define IPCC_C2SCR			U(0x18)
+#define IPPC_C2TOC1SR			U(0x1C)
+
+/* Offset within a core instance */
+#define IPCC_CR				U(0x0)
+#define IPCC_MR				U(0x4)
+#define IPCC_SCR			U(0x8)
+#define IPCC_TOSR			U(0xC)
+#define IPCC_SECCFGR			U(0x80)
+#define IPCC_PRIVCFGR			U(0x84)
+#define IPCC_CIDCFGR			U(0x98)
+
+/* Mask for  channel rxo and txf */
+#define IPCC_ALL_MR_TXF_CH_MASK		GENMASK_32(31, 16)
+#define IPCC_ALL_MR_RXO_CH_MASK		GENMASK_32(15, 0)
+#define IPCC_ALL_SR_CH_MASK		GENMASK_32(15, 0)
+
+/* Define for core instance register */
+/* PROCESSOR  RECEIVE CHANNEL OCCUPIED NON-SECURE INTERRUPT ENABLE */
+#define IPCC_CR_RXOIE			BIT(0)
+/* PROCESSOR  RECEIVE CHANNEL OCCUPIED SECURE INTERRUPT ENABLE */
+#define IPCC_CR_SECRXOIE		BIT(1)
+/* PROCESSOR TRANSMIT CHANNEL FREE NON-SECURE INTERRUPT ENABLE */
+#define IPCC_CR_TXFIE			BIT(16)
+ /* PROCESSOR TRANSMIT CHANNEL FREE SECURE INTERRUPT ENABLE */
+#define IPCC_CR_SECTXFIE		BIT(17)
+
+#define IPCC_MR_CH1FM_POS		U(16)
+#define IPCC_SCR_CH1S_POS		U(16)
+
 /*
  * CIDCFGR register bitfields
  */
@@ -58,6 +98,24 @@
 
 #define IPCC_NB_MAX_CID_SUPPORTED	U(7)
 
+enum {
+	IPCC_ITR_RXO,
+	IPCC_ITR_TXF,
+	IPCC_ITR_NUM,
+};
+
+/* IPCC mbox data */
+struct stm32_ipcc_mbx_data {
+	/* Lock to protect concurrent access */
+	unsigned int lock;
+	uint32_t channel_sec_mask;
+	uint32_t channel_enable_mask;
+	unsigned int num_ch;
+	mbox_rx_callback_t rx_cb;
+	mbox_tx_callback_t tx_cb;
+	struct mbox_dev *cb_ctx;
+};
+
 struct ipcc_pdata {
 	/*
 	 * An IPCC has nb_channels_cfg channel configuration for its
@@ -68,6 +126,16 @@ struct ipcc_pdata {
 	vaddr_t base;
 	struct rif_conf_data conf_data;
 
+	/* Remote proc base address */
+	vaddr_t rbase;
+	/* Local proc base address */
+	vaddr_t lbase;
+	struct itr_chip *itr_chip[IPCC_ITR_NUM];
+	size_t itr_num[IPCC_ITR_NUM];
+	struct itr_handler *itr[IPCC_ITR_NUM];
+
+	/* Single mailbox user (mailbox framework) */
+	struct stm32_ipcc_mbx_data data;
 	STAILQ_ENTRY(ipcc_pdata) link;
 };
 
@@ -231,6 +299,306 @@ stm32_ipcc_pm(enum pm_op op, unsigned int pm_hint,
 	return TEE_SUCCESS;
 }
 
+static bool ipcc_channel_is_active(vaddr_t lbase, unsigned int chn)
+{
+	return io_read32(lbase + IPCC_TOSR) & BIT(chn);
+}
+
+static void ipcc_channel_transmit(vaddr_t lbase, unsigned int chn, bool enable)
+{
+	if (enable)
+		io_clrbits32(lbase + IPCC_MR, BIT(chn + IPCC_MR_CH1FM_POS));
+	else
+		io_setbits32(lbase + IPCC_MR, BIT(chn + IPCC_MR_CH1FM_POS));
+}
+
+static void ipcc_channel_receive(vaddr_t lbase, unsigned int chn, bool enable)
+{
+	if (enable)
+		io_clrbits32(lbase + IPCC_MR, BIT(chn));
+	else
+		io_setbits32(lbase + IPCC_MR, BIT(chn));
+}
+
+static enum itr_return stm32_ipcc_mbox_rxo_isr(struct itr_handler *h)
+{
+	struct ipcc_pdata *pdata = h->data;
+	struct stm32_ipcc_mbx_data *data = &pdata->data;
+	unsigned int value = 0;
+	uint32_t mask = 0;
+	unsigned int i = 0;
+
+	mask = ~io_read32(pdata->lbase + IPCC_MR) & IPCC_ALL_MR_RXO_CH_MASK;
+	mask &= io_read32(pdata->rbase + IPCC_TOSR) & IPCC_ALL_SR_CH_MASK;
+
+	/* Get mask for enabled channels only */
+	mask &= data->channel_enable_mask;
+
+	for (i = 0; i < data->num_ch; i++) {
+		if (!(BIT(i) & mask))
+			continue;
+		FMSG("rx channel = %"PRIu32, i);
+		if (data->rx_cb)
+			data->rx_cb(data->cb_ctx, i, &value);
+		/* Clear rxo flag */
+		io_write32(pdata->lbase + IPCC_SCR, BIT(i));
+	}
+	return ITRR_HANDLED;
+}
+
+static enum itr_return stm32_ipcc_mbox_txf_isr(struct itr_handler *h)
+{
+	struct ipcc_pdata *pdata = h->data;
+	struct stm32_ipcc_mbx_data *data = &pdata->data;
+	uint32_t mask = 0;
+	uint32_t i = 0;
+
+	mask = ~io_read32(pdata->lbase + IPCC_MR) & IPCC_ALL_MR_TXF_CH_MASK;
+	mask = mask >> IPCC_MR_CH1FM_POS;
+
+	mask &= ~io_read32(pdata->lbase + IPCC_TOSR) & IPCC_ALL_SR_CH_MASK;
+
+	/* Get mask for enabled channels only */
+	mask &= data->channel_enable_mask;
+
+	for (i = 0; i < data->num_ch; i++) {
+		if (BIT(i) & mask) {
+			if (data->tx_cb)
+				data->tx_cb(data->cb_ctx, i);
+			/* Disable txf interrupt */
+			ipcc_channel_transmit(pdata->lbase, i, false);
+		}
+	}
+	return ITRR_HANDLED;
+}
+
+static TEE_Result stm32_ipcc_mbox_send(const struct mbox_desc *desc,
+				       unsigned int id,
+				       const void *buff __unused, size_t size)
+{
+	struct ipcc_pdata *pdata = (struct ipcc_pdata *)desc->priv;
+	struct stm32_ipcc_mbx_data *data = &pdata->data;
+
+	/* No data transmission, only doorbell */
+	if (size)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	if (id >= data->num_ch || !(BIT(id) & data->channel_sec_mask)) {
+		EMSG("invalid id %u", id);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	FMSG("Send msg on channel %u", id);
+
+	/* Check that the channel is free */
+	if (ipcc_channel_is_active(pdata->lbase, id)) {
+		FMSG("Waiting for channel to be freed");
+		return TEE_ERROR_BUSY;
+	}
+	/* Set channel txs Flag */
+	io_write32(pdata->lbase + IPCC_SCR, BIT(id + IPCC_SCR_CH1S_POS));
+	/* Enable channel txf interrupt */
+	ipcc_channel_transmit(pdata->lbase, id, true);
+
+	return TEE_SUCCESS;
+}
+
+static size_t stm32_ipcc_mbox_get_mtu(const struct mbox_desc *desc __unused)
+{
+	/* No data transfer capability */
+	return 0;
+}
+
+static size_t stm32_ipcc_mbox_channel_count(const struct mbox_desc *desc)
+{
+	struct ipcc_pdata *pdata = (struct ipcc_pdata *)desc->priv;
+	struct stm32_ipcc_mbx_data *data = &pdata->data;
+
+	assert(data->num_ch);
+
+	return data->num_ch;
+}
+
+static TEE_Result stm32_ipcc_mbox_register_callback(const
+						    struct mbox_desc *desc,
+						    mbox_rx_callback_t rx_cb,
+						    mbox_tx_callback_t tx_cb,
+						    struct mbox_dev *cb_ctx)
+{
+	struct ipcc_pdata *pdata = (struct ipcc_pdata *)desc->priv;
+	struct stm32_ipcc_mbx_data *data = &pdata->data;
+	uint32_t except = 0;
+	TEE_Result res = TEE_SUCCESS;
+
+	except = cpu_spin_lock_xsave(&data->lock);
+	if (!data->rx_cb && !data->tx_cb && !data->cb_ctx) {
+		data->rx_cb = rx_cb;
+		data->tx_cb = tx_cb;
+		data->cb_ctx = cb_ctx;
+		res = TEE_SUCCESS;
+	} else {
+		res = TEE_ERROR_BAD_STATE;
+	}
+	cpu_spin_unlock_xrestore(&data->lock, except);
+
+	return res;
+}
+
+static TEE_Result stm32_ipcc_mbox_enable(const struct mbox_desc *desc,
+					 bool enable, unsigned int id)
+{
+	struct ipcc_pdata *pdata = (struct ipcc_pdata *)desc->priv;
+	struct stm32_ipcc_mbx_data *data = &pdata->data;
+	uint32_t channel_enable_mask = 0;
+	uint32_t except = 0;
+
+	if (!(BIT(id) & data->channel_sec_mask))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	except = cpu_spin_lock_xsave(&data->lock);
+
+	if (enable)
+		channel_enable_mask = (BIT(id) & data->channel_sec_mask) |
+			data->channel_enable_mask;
+	else
+		channel_enable_mask = ~(BIT(id) & data->channel_sec_mask) &
+			data->channel_enable_mask;
+
+	if (channel_enable_mask && !data->channel_enable_mask) {
+		/* Enable secure txf and rxo interrupt */
+		io_setbits32(pdata->lbase + IPCC_CR,
+			     IPCC_CR_SECTXFIE | IPCC_CR_SECRXOIE);
+	}
+
+	if (!channel_enable_mask && data->channel_enable_mask) {
+		/* Disable secure txf and rxo interrupt */
+		io_clrbits32(pdata->lbase + IPCC_CR,
+			     IPCC_CR_SECTXFIE | IPCC_CR_SECRXOIE);
+	}
+
+	/*
+	 * Update mask and then enable rxo interrupt
+	 * since mask is used within rxo interrupt handler and
+	 * do the opposite for disable
+	 */
+	if (enable)
+		data->channel_enable_mask = channel_enable_mask;
+	ipcc_channel_receive(pdata->lbase, id, enable);
+	if (!enable)
+		data->channel_enable_mask = channel_enable_mask;
+
+	cpu_spin_unlock_xrestore(&data->lock, except);
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result stm32_ipcc_mbox_complete(const struct mbox_desc *desc,
+					   unsigned int id)
+{
+	struct ipcc_pdata *pdata = (struct ipcc_pdata *)desc->priv;
+	struct stm32_ipcc_mbx_data *data = &pdata->data;
+
+	if (!(BIT(id) & data->channel_enable_mask))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return TEE_SUCCESS;
+}
+
+static const struct mbox_ops stm32_ipcc_mbox_ops = {
+	.send = stm32_ipcc_mbox_send,
+	.register_callback = stm32_ipcc_mbox_register_callback,
+	.max_data_size = stm32_ipcc_mbox_get_mtu,
+	.max_channel_count = stm32_ipcc_mbox_channel_count,
+	.channel_interrupt = stm32_ipcc_mbox_enable,
+	.complete = stm32_ipcc_mbox_complete,
+};
+
+static void stm32_ipcc_mbox_init(struct ipcc_pdata *pdata, const void *fdt,
+				 int node)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	uint32_t i = 0;
+	struct stm32_ipcc_mbx_data *data = &pdata->data;
+	struct mbox_desc *desc = NULL;
+
+	/* Null phandle means no client of this node, no mailbox support */
+	if (!fdt_get_phandle(fdt, node))
+		return;
+
+	res = interrupt_dt_get_by_name(fdt, node, "rx",
+				       &pdata->itr_chip[IPCC_ITR_RXO],
+				       &pdata->itr_num[IPCC_ITR_RXO]);
+	if (res)
+		panic();
+	res = interrupt_dt_get_by_name(fdt, node, "tx",
+				       &pdata->itr_chip[IPCC_ITR_TXF],
+				       &pdata->itr_num[IPCC_ITR_TXF]);
+	if (res)
+		panic();
+
+	desc = calloc(1, sizeof(*desc));
+	if (!desc)
+		panic();
+
+	desc->ops = &stm32_ipcc_mbox_ops;
+	desc->priv = pdata;
+
+	/*
+	 * IPCC Clock is enable for product life, since clock is required for
+	 * all mailbox framework users relying on IPCC
+	 */
+	if (clk_enable(pdata->ipcc_clock))
+		panic("Cannot access ipcc clock");
+
+	/* Disable secure rxo and txf interrupts */
+	io_clrbits32(pdata->lbase + IPCC_CR, IPCC_CR_SECTXFIE |
+		     IPCC_CR_SECRXOIE);
+
+	/* Fill max Channel */
+	data->num_ch = io_read32(pdata->base + IPCC_HWCFGR) &
+		IPCC_HWCFGR_CHAN_MASK;
+
+	/* Set channel_sec_mask according to rif protection */
+	data->channel_sec_mask = io_read32(pdata->lbase + IPCC_SECCFGR);
+	data->channel_enable_mask = 0;
+
+	/* Fix Me : Possibly Add Check on cid filtering and privileged */
+
+	for (i = 0; i < data->num_ch; i++) {
+		/* Clear rxo status */
+		io_write32(pdata->lbase + IPCC_SCR, BIT(i));
+		/* Mask channel rxo and txf interrupts */
+		ipcc_channel_receive(pdata->lbase, i, false);
+		ipcc_channel_transmit(pdata->lbase, i, false);
+	}
+	/* Register irq handler */
+	res = interrupt_alloc_add_handler(pdata->itr_chip[IPCC_ITR_RXO],
+					  pdata->itr_num[IPCC_ITR_RXO],
+					  stm32_ipcc_mbox_rxo_isr,
+					  ITRF_TRIGGER_LEVEL, pdata,
+					  &pdata->itr[IPCC_ITR_RXO]);
+	if (res)
+		panic("Cannot register ipcc rx handler");
+
+	res = interrupt_alloc_add_handler(pdata->itr_chip[IPCC_ITR_TXF],
+					  pdata->itr_num[IPCC_ITR_TXF],
+					  stm32_ipcc_mbox_txf_isr,
+					  ITRF_TRIGGER_LEVEL, pdata,
+					  &pdata->itr[IPCC_ITR_TXF]);
+	if (res)
+
+		panic("Cannot register ipcc tx handler");
+
+	interrupt_enable(pdata->itr_chip[IPCC_ITR_RXO],
+			 pdata->itr[IPCC_ITR_RXO]->it);
+	interrupt_enable(pdata->itr_chip[IPCC_ITR_TXF],
+			 pdata->itr[IPCC_ITR_TXF]->it);
+
+	/* Register device to mailbox framework */
+	if (mbox_dt_register(fdt, node, desc))
+		panic("Failed to register ipcc mbox");
+}
+
 static TEE_Result parse_dt(const void *fdt, int node, struct ipcc_pdata *ipcc_d)
 {
 	int lenp = 0;
@@ -240,6 +608,7 @@ static TEE_Result parse_dt(const void *fdt, int node, struct ipcc_pdata *ipcc_d)
 	const fdt32_t *cuint = NULL;
 	struct dt_node_info info = { };
 	struct io_pa_va addr = { };
+	uint32_t proc_id = 0;
 
 	fdt_fill_device_info(fdt, &info, node);
 	assert(info.reg != DT_INFO_INVALID_REG &&
@@ -248,6 +617,18 @@ static TEE_Result parse_dt(const void *fdt, int node, struct ipcc_pdata *ipcc_d)
 	addr.pa = info.reg;
 	ipcc_d->base = io_pa_or_va_secure(&addr, info.reg_size);
 	assert(ipcc_d->base);
+	/*
+	 * Field st-proc-id, fixes CPU core on which driver is instantiated
+	 * according to proc ID mailbox lbase. rbase mailbox are swapped.
+	 */
+	(void)fdt_read_uint32(fdt, node, "st,proc-id", &proc_id);
+	if (proc_id == 2) {
+		ipcc_d->lbase = ipcc_d->base + IPCC_C2CR;
+		ipcc_d->rbase = ipcc_d->base + IPCC_C1CR;
+	} else {
+		ipcc_d->lbase = ipcc_d->base + IPCC_C1CR;
+		ipcc_d->rbase = ipcc_d->base + IPCC_C2CR;
+	}
 
 	/* Gate the IP */
 	res = clk_dt_get_by_index(fdt, node, 0, &ipcc_d->ipcc_clock);
@@ -312,7 +693,10 @@ static TEE_Result stm32_ipcc_probe(const void *fdt, int node,
 
 	register_pm_core_service_cb(stm32_ipcc_pm, ipcc_d, "stm32-ipcc");
 
-	return res;
+	if (IS_ENABLED(CFG_DRIVERS_MAILBOX))
+		stm32_ipcc_mbox_init(ipcc_d, fdt, node);
+
+	return TEE_SUCCESS;
 
 err:
 	/* Free all allocated resources */
-- 
2.47.2

